### PR TITLE
Reset password endpoints

### PR DESCRIPTION
**Pull request by MustafaOnurKilinc**

### CODE REVIEW

File: config/default.json [hunk 1]
```diff
@@ -18,5 +18,16 @@
...
     "operatorsAliases": true
+  },
+  "mailer": {
+    "defaultFrom": "kommunitybackend@gmail.com",
+    "pool": true,
+    "host": "smtp.gmail.com",
+    "port": 465,
+    "secure": true,
+    "auth": {
+      "user": "kommunitybackend@gmail.com",
```
Comment (SelmanKahya) [inline config/default.json]: lets get these from envrionment instead, we can define them in heroku and pass it inruntime

Comment (SelmanKahya) [reply config/default.json]: it is always better practice to not commit your important keys/passwds

File: src/lib/app.js [hunk 2]
```diff
@@ -79,9 +83,13 @@ export default class App {
     });
   };
 
+  initMailer = (): void => {
```
Comment (SelmanKahya) [inline src/lib/app.js]: lets call this func initClients

File: src/lib/app.js [hunk 3]
```diff
@@ -79,9 +83,13 @@ export default class App {
...
+    this.mailer = Nodemailer.createTransport(this.config.mailer);
```
Comment (SelmanKahya) [inline src/lib/app.js]: lets do `this.clients.mailer`

File: src/lib/app.js [hunk 4]
```diff
@@ -91,19 +99,20 @@ export default class App {
     this.express.set('views', path.join(__dirname, viewFolderPath));
 
     if (sentryConfig) {
+      this.sentry = Sentry;
       const {
         dsn, debug, environment, sampleRate, attachStacktrace,
       } = sentryConfig;
       // Error Handling
-      sentry.init({
+      this.sentry.init({
```
Comment (SelmanKahya) [inline src/lib/app.js]: lets do `this.clients.sentry`

File: src/lib/app.js [hunk 5]
```diff
@@ -212,9 +221,17 @@ export default class App {
...
     });
-  }
+  };
 
   registerRoute = (route: string, router: express$Router): void => {
     this.express.use(route, router);
-  }
+  };
+
+  sendMail = (messageOptions: AppMailMessage): void => {
```
Comment (SelmanKahya) [inline src/lib/app.js]: lets create `src/clients/mailer.js` and call it

File: src/lib/app.js [hunk 6]
```diff
@@ -212,9 +221,17 @@ export default class App {
...
+    const message = messageOptions;
+    message.from = this.config.mailer.defaultFrom;
+    this.mailer.sendMail(message, (error) => {
```
Comment (SelmanKahya) [inline src/lib/app.js]: in the long term, we create create a queue for sending emails

Comment (SelmanKahya) [reply src/lib/app.js]: see https://github.com/OptimalBits/bull

File: src/models/user.js [hunk 7]
```diff
@@ -28,6 +28,13 @@ module.exports = (sequelize: Sequelize, dataTypes: DataTypes) => {
         len: [32, 32],
       },
     },
+    passwordResetUuid: {
+      type: dataTypes.UUID,
+      field: 'password_reset_uuid',
```
Comment (SelmanKahya) [inline src/models/user.js]: we need to make sure it expires after certain amount of time, and when user uses it once, we should invalidate

File: src/routes/api/member.js [hunk 8]
```diff
@@ -42,6 +42,62 @@ const routes = (app: App): void => {
     res.json({ success: true });
   });
 
+  router.post('/forgot-password', (req: exExpress$Request, res: express$Response) => {
```
Comment (SelmanKahya) [inline src/routes/api/member.js]: as discussed over discord, we can move all these to graphql

File: src/routes/api/member.js [hunk 9]
```diff
@@ -42,6 +42,62 @@ const routes = (app: App): void => {
...
+    const { email } = req.body;
```
Comment (SelmanKahya) [inline src/routes/api/member.js]: we should get the email from user object (that we authenticated)

File: src/routes/api/member.js [hunk 10]
```diff
@@ -42,6 +42,62 @@ const routes = (app: App): void => {
...
+    // TODO handle error, transform response
+    const passwordResetUuid = uuid();
+    app.models.User.update(
+      { passwordResetUuid },
+      { where: { email } },
+    ).then((rowsUpdated: number) => {
+      if (rowsUpdated > 0) {
+        app.sendMail({
```
Comment (SelmanKahya) [inline src/routes/api/member.js]: lets integrate with this:
https://github.com/yads/nodemailer-express-handlebars

so we can send emails with html easily

File: src/routes/api/member.js [hunk 11]
```diff
@@ -42,6 +42,62 @@ const routes = (app: App): void => {
...
+      // always return 200.
+      res.statusCode = 200;
+      res.send();
+    }).catch((err) => {
+      res.json({ err });
+    });
+  });
+
+  router.get('/reset-password', (req: exExpress$Request, res: express$Response) => {
+    const { guid } = req.query;
```
Comment (SelmanKahya) [inline src/routes/api/member.js]: what does guid stand for? acilimi nedir?

File: src/routes/api/member.js [hunk 12]
```diff
@@ -42,6 +42,62 @@ const routes = (app: App): void => {
...
+    app.models.User.findOne({
+      where: { passwordResetUuid: guid },
```
Comment (SelmanKahya) [inline src/routes/api/member.js]: we should use jwt to create reset password tokens:
https://www.smashingmagazine.com/2017/11/safe-password-resets-with-json-web-tokens/

File: src/routes/api/member.js [hunk 13]
```diff
@@ -42,6 +42,62 @@ const routes = (app: App): void => {
...
+      }
+      // always return 200.
+      res.statusCode = 200;
+      res.send();
+    }).catch((err) => {
+      res.json({ err });
+    });
+  });
+
+  router.get('/reset-password', (req: exExpress$Request, res: express$Response) => {
```
Comment (SelmanKahya) [inline src/routes/api/member.js]: why do we need get reset-password route in api?

File: config/default.json (hunk 1 above)
Comment (MustafaOnurKilinc) [reply config/default.json]: Ok.

File: src/lib/app.js (hunk 3 above)
Comment (MustafaOnurKilinc) [reply src/lib/app.js]: Ok.

File: src/lib/app.js (hunk 4 above)
Comment (MustafaOnurKilinc) [reply src/lib/app.js]: Ok.

File: src/routes/api/member.js (hunk 11 above)
Comment (MustafaOnurKilinc) [reply src/routes/api/member.js]: uuid yi .net kodlarken guid diye kullanıyoruz. oradan kalma bir alışkanlık.

File: src/routes/api/member.js (hunk 13 above)
Comment (MustafaOnurKilinc) [reply src/routes/api/member.js]: email linkinden geri dönüş için gerekli.

File: src/lib/app.js (hunk 5 above)
Comment (MustafaOnurKilinc) [reply src/lib/app.js]: Ok.

File: src/models/user.js (hunk 7 above)
Comment (MustafaOnurKilinc) [reply src/models/user.js]: Ok. We use jwt token.

File: src/routes/api/member.js (hunk 8 above)
Comment (MustafaOnurKilinc) [reply src/routes/api/member.js]: Ok.

File: src/routes/api/member.js (hunk 11 above)
Comment (MustafaOnurKilinc) [reply src/routes/api/member.js]: uuid in .net deki karşılığı. Alışkanlık.

Comment (SelmanKahya) [reply src/routes/api/member.js]: uuid kullanalim istersen, consistent olsun

Comment (MustafaOnurKilinc) [reply src/routes/api/member.js]: Ok.

File: src/lib/app.js (hunk 6 above)
Comment (MustafaOnurKilinc) [reply src/lib/app.js]: Ok.

File: src/routes/api/member.js (hunk 10 above)
Comment (MustafaOnurKilinc) [reply src/routes/api/member.js]: Ok.

File: src/routes/api/member.js (hunk 9 above)
Comment (MustafaOnurKilinc) [reply src/routes/api/member.js]: Bu ilk adımda, kullanıcı parolasını sıfırlamak için email adresini form ile gönderiyor. Henüz oturum bulunmadığından user nesnesinden alamayız. Yanılıyor muyum?

Comment (SelmanKahya) [reply src/routes/api/member.js]: dogru, fakat emailden kullaniciyi bulmamiz lazim, yoksa mail gondermemeli.

**Comment by SelmanKahya** [comment]
any updates on this?

**Pull request closed without merging**